File: 567. Permutation in String/Solution.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int n = s1.length(), m = s2.length();
        if (m < n) return false;
        int[] freq = new int[26];
        for (int i = 0; i < n; i++) {
            freq[s1.charAt(i) - 'a']++;
            freq[s2.charAt(i) - 'a']--;
        }
        if (check(freq)) return true;
        for (int i = n; i < m; i++) {
            freq[s2.charAt(i) - 'a']--;
            freq[s2.charAt(i - n) - 'a']++;
            if (check(freq)) return true;
        }
        return false;
    }

    public boolean check(int[] freq) {
        int i;
        for (i = 0; i < 26 && freq[i] == 0; i++) { }
        return i == 26;
    }
}

File: 567. Permutation in String/Solution.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n, m = len(s1), len(s2)
        if m < n: return False
        freq = [0 for i in range(26)]
        for i in range(n):
            freq[self.getIdx(s1[i])] += 1
            freq[self.getIdx(s2[i])] -= 1
        if self.check(freq): return True
        for i in range(n, m):
            freq[self.getIdx(s2[i])] -= 1
            freq[self.getIdx(s2[i - n])] += 1
            if self.check(freq): return True
        return False

    def getIdx(self, c):
        return ord(c) - ord('a')

    def check(self, freq):
        for i in range(26):
            if freq[i] != 0: return False
        return True